Sprint 12

kover opened the pull request.

kover committed:
Define data models

kover committed:
Define ImageListService with protocol

kover committed:
Set date decoding strategy

kover committed:
Inject service into ImagesListViewController

kover committed:
Load images feed from network

kover committed:
Toggle like operations

kover committed:
Show sintle image

kover committed:
Implement logout feature

kover committed:
Cleanup and cell configuration optimization

Add logout confirmation

andrew-wgt commented on Imaginarium/ImagesList/ImagesListCell.swift:
```diff
@@ -45,17 +42,36 @@ final class ImagesListCell: UITableViewCell {
     @IBOutlet private var gradientView: UIView!
```
Молодец, что указал private для аутлетов! Также хорошей практикой является указание модификатора weak к IBOutlet. Вероятность retain cycle тут мала, но и нам ничего не стоит написать слабые ссылки)

kover replied:
Спасибо, добавлю.

andrew-wgt commented on Imaginarium/ImagesList/ImagesListViewController.swift:
```diff
@@ -57,10 +76,15 @@ extension ImagesListViewController: UITableViewDataSource {
+            postedAt = dateFormatter.string(from: createdAt)
+        }
+        imageListCell.configureCell(forPhoto: photo.thumbImageURL, postedAt: postedAt, liked: photo.isLiked) {
+            tableView.reloadRows(at: [indexPath], with: .automatic)
```
Эта строчка избыточна, так как ячейка и так обновится в configCell. Можно убрать комплишн

kover replied:
Понял

andrew-wgt commented on Imaginarium/ImagesList/ImagesListViewController.swift:
```diff
@@ -74,15 +98,65 @@ extension ImagesListViewController: UITableViewDelegate {
+                cell.setIsLiked(isLiked: !photo.isLiked)
+                UIBlockingProgressHUD.dismiss()
+            case .failure:
+                UIBlockingProgressHUD.dismiss()
```
UIBlockingProgressHUD можно вывести за пределы switch. Так как его вызывать в любом случае, а дублировать строчку не придётся)

kover replied:
Спасибо!

andrew-wgt commented on Imaginarium/ImagesList/ImagesListViewController.swift:
```diff
@@ -74,15 +98,65 @@ extension ImagesListViewController: UITableViewDelegate {
+        }
+        let photo  = photos[indexPath.row]
+        UIBlockingProgressHUD.show()
+        imagesListService?.changeLike(photoId: photo.id, isLike: !photo.isLiked, { result in
```
😱 Забыт [weak self]

kover replied:
Поправил :)

andrew-wgt commented on Imaginarium/Profile/ProfileViewController.swift:
```diff
@@ -56,7 +59,46 @@ final class ProfileViewController: UIViewController {
+    @objc private func logoutTap() {
+        let alert = UIAlertController(
+            title: "Пока, пока!",
+            message: "Уверены что хотите выйте?",
```
Опечатка)

kover replied:
:) спасибо

andrew-wgt commented on Imaginarium/ServiceAccessLayer/UnsplashAPI/ImagesListService.swift:
```diff
@@ -0,0 +1,110 @@
+                    )
+                lastLoadedPage = nextPage
+            case .failure:
+                // TODO: Handle and display error
```
Ошибку всё таки стоит обработать

kover replied:
Сделал

kover committed:
Review changes

- Add weak self in imageListCellDidTapLike
- Handle load images page errors
- Fix typo in logout confirmation
- Other fixes and small refactor
